sdks/wasm/zlib-helper.c

```diff
@@ -0,0 +1,287 @@
+#include <emscripten.h>
+#include <string.h>
+
+#include <mono/jit/jit.h>
+
+#include "zlib.h"
+
+#ifndef MONO_API
+#define MONO_API
+#endif
+
+#ifndef TRUE
+#define FALSE 0
+#define TRUE 1
+#endif
+
+/*
+ * Basic data types
+ */
+typedef int            gint;
+typedef unsigned int   guint;
+typedef short          gshort;
+typedef unsigned short gushort;
+typedef long           glong;
+typedef unsigned long  gulong;
+typedef void *         gpointer;
+typedef const void *   gconstpointer;
+typedef char           gchar;
+typedef unsigned char  guchar;
+
+/* Types defined in terms of the stdint.h */
+typedef int8_t         gint8;
+typedef uint8_t        guint8;
+typedef int16_t        gint16;
+typedef uint16_t       guint16;
+typedef int32_t        gint32;
+typedef uint32_t       guint32;
+typedef int64_t        gint64;
+typedef uint64_t       guint64;
+typedef float          gfloat;
+typedef double         gdouble;
+typedef int32_t        gboolean;
+
+#define BUFFER_SIZE 4096
+#define ARGUMENT_ERROR -10
+#define IO_ERROR -11
+#define MONO_EXCEPTION -12
+
+#define z_malloc(size)          ((gpointer) malloc(size))
+#define z_malloc0(size)         ((gpointer) calloc(1,size))
+#define z_new(type,size)        ((type *) z_malloc (sizeof (type) * (size)))
+#define z_new0(type,size)       ((type *) z_malloc0 (sizeof (type)* (size)))
+
+typedef gint (*read_write_func) (guchar *buffer, gint length, void *gchandle);
+struct _ZStream {
+	z_stream *stream;
+	guchar *buffer;
+	read_write_func func;
+	void *gchandle;
+	guchar compress;
+	guchar eof;
+	guint32 total_in;
+};
+typedef struct _ZStream ZStream;
+
+MONO_API ZStream *CreateZStream (gint compress, guchar gzip, read_write_func func, void *gchandle);
+MONO_API gint CloseZStream (ZStream *zstream);
+MONO_API gint Flush (ZStream *stream);
+MONO_API gint ReadZStream (ZStream *stream, guchar *buffer, gint length);
+MONO_API gint WriteZStream (ZStream *stream, guchar *buffer, gint length);
+static gint flush_internal (ZStream *stream, gboolean is_final);
+
+void zlibhelper_initialize_internals ()
+{
+	mono_add_internal_call ("System.IO.Compression.DeflateStreamNative::CreateZStream", CreateZStream);
+	mono_add_internal_call ("System.IO.Compression.DeflateStreamNative::CloseZStream", CloseZStream);
```

Comment:
Wouldn't be possible to implement these as normal c# icalls ? Then the linker can link them out.

Reply:
See next commit.